From 9cab33158cc21f0403b3f0603718f26c8364a339 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Wed, 7 Sep 2011 16:29:26 -0600
Subject: [PATCH] Send an email when change moves to a new branch

Add an Moved.vm email template and use it when a change
is moved to a new branch to send email to reviewers,
watchers...

Change-Id: Ia9c7f0bbd00eec34326d458fbaefd0827f2885d5
---
 Documentation/config-mail.txt                      |    7 +++
 .../gerrit/server/changedetail/MoveChange.java     |   13 +++++-
 .../gerrit/server/config/GerritRequestModule.java  |    2 +
 .../gerrit/server/mail/ChangeMovedSender.java      |   48 ++++++++++++++++++++
 .../com/google/gerrit/server/mail/ChangeMoved.vm   |   44 ++++++++++++++++++
 5 files changed, 113 insertions(+), 1 deletions(-)
 create mode 100644 gerrit-server/src/main/java/com/google/gerrit/server/mail/ChangeMovedSender.java
 create mode 100644 gerrit-server/src/main/resources/com/google/gerrit/server/mail/ChangeMoved.vm

diff --git a/Documentation/config-mail.txt b/Documentation/config-mail.txt
index 8aa7d08..7155a4e 100644
--- a/Documentation/config-mail.txt
+++ b/Documentation/config-mail.txt
@@ -32,6 +32,13 @@ The `Abandoned.vm` template will determine the contents of the email related
 to a change being abandoned.  It is a `ChangeEmail`: see `ChangeSubject.vm` and
 `ChangeFooter.vm`.
 
+Moved.vm
+~~~~~~~~
+
+The `Moved.vm` template will determine the contents of the email related
+to a change being moved to another destination branch.  It
+is a `ChangeEmail`: see `ChangeSubject.vm` and `ChangeFooter.vm`.
+
 ChangeFooter.vm
 ~~~~~~~~~~~~~~~
 
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/changedetail/MoveChange.java b/gerrit-server/src/main/java/com/google/gerrit/server/changedetail/MoveChange.java
index 78daab5..075e238 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/changedetail/MoveChange.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/changedetail/MoveChange.java
@@ -23,6 +23,7 @@ import com.google.gerrit.reviewdb.ReviewDb;
 import com.google.gerrit.server.ApprovalsUtil;
 import com.google.gerrit.server.ChangeUtil;
 import com.google.gerrit.server.IdentifiedUser;
+import com.google.gerrit.server.mail.ChangeMovedSender;
 import com.google.gerrit.server.mail.EmailException;
 import com.google.gerrit.server.project.InvalidChangeOperationException;
 import com.google.gerrit.server.project.NoSuchChangeException;
@@ -48,6 +49,7 @@ public class MoveChange implements Callable<VoidResult> {
   private final ChangeControl.Factory changeControlFactory;
   private final ReviewDb db;
   private final IdentifiedUser currentUser;
+  private final ChangeMovedSender.Factory senderFactory;
 
   private final PatchSet.Id patchSetId;
   private final String branch;
@@ -56,12 +58,15 @@ public class MoveChange implements Callable<VoidResult> {
 
   @Inject
   MoveChange(final ChangeControl.Factory changeControlFactory, final ReviewDb db,
-      final IdentifiedUser currentUser, final @Assisted PatchSet.Id patchSetId,
+      final IdentifiedUser currentUser,
+      final ChangeMovedSender.Factory senderFactory, 
+      final @Assisted PatchSet.Id patchSetId,
       final @Assisted("branch") String branch,
       final @Assisted("message") @Nullable String changeComment) {
     this.changeControlFactory = changeControlFactory;
     this.db = db;
     this.currentUser = currentUser;
+    this.senderFactory = senderFactory;
 
     this.patchSetId = patchSetId;
     this.branch = branch;
@@ -113,6 +118,12 @@ public class MoveChange implements Callable<VoidResult> {
 
     ApprovalsUtil.clearCurrentPatchSetApprovals(db, updatedChange);
 
+    // Email the reviewers
+    final ChangeMovedSender cm = senderFactory.create(updatedChange);
+    cm.setFrom(currentUser.getAccountId());
+    cm.setChangeMessage(cmsg);
+    cm.send();
+
     return VoidResult.INSTANCE;
   }
 }
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/config/GerritRequestModule.java b/gerrit-server/src/main/java/com/google/gerrit/server/config/GerritRequestModule.java
index 26a33b6..85fb764 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/config/GerritRequestModule.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/config/GerritRequestModule.java
@@ -39,6 +39,7 @@ import com.google.gerrit.server.git.ReceiveCommits;
 import com.google.gerrit.server.git.SubmoduleOp;
 import com.google.gerrit.server.mail.AbandonedSender;
 import com.google.gerrit.server.mail.AddReviewerSender;
+import com.google.gerrit.server.mail.ChangeMovedSender;
 import com.google.gerrit.server.mail.CommentSender;
 import com.google.gerrit.server.mail.CreateChangeSender;
 import com.google.gerrit.server.mail.MergeFailSender;
@@ -94,6 +95,7 @@ public class GerritRequestModule extends FactoryModule {
     factory(PublishDraft.Factory.class);
     factory(ReplacePatchSetSender.Factory.class);
     factory(AbandonedSender.Factory.class);
+    factory(ChangeMovedSender.Factory.class);
     factory(RemoveReviewer.Factory.class);
     factory(RestoreChange.Factory.class);
     factory(RestoredSender.Factory.class);
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/mail/ChangeMovedSender.java b/gerrit-server/src/main/java/com/google/gerrit/server/mail/ChangeMovedSender.java
new file mode 100644
index 0000000..817e96a
--- /dev/null
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/mail/ChangeMovedSender.java
@@ -0,0 +1,48 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.gerrit.server.mail;
+
+import com.google.gerrit.reviewdb.Change;
+import com.google.gerrit.server.config.AnonymousCowardName;
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+
+/** Send notice about a change being move to a different destination branch. */
+public class ChangeMovedSender extends ReplyToChangeSender {
+  public static interface Factory extends
+      ReplyToChangeSender.Factory<ChangeMovedSender> {
+    ChangeMovedSender create(Change change);
+  }
+
+  @Inject
+  public ChangeMovedSender(EmailArguments ea,
+      @AnonymousCowardName String anonymousCowardName, @Assisted Change c) {
+    super(ea, anonymousCowardName, c, "change moved");
+  }
+
+  @Override
+  protected void init() throws EmailException {
+    super.init();
+
+    ccAllApprovals();
+    bccStarredBy();
+    bccWatchesNotifyAllComments();
+  }
+
+  @Override
+  protected void formatChange() throws EmailException {
+    appendText(velocifyFile("ChangeMoved.vm"));
+  }
+}
diff --git a/gerrit-server/src/main/resources/com/google/gerrit/server/mail/ChangeMoved.vm b/gerrit-server/src/main/resources/com/google/gerrit/server/mail/ChangeMoved.vm
new file mode 100644
index 0000000..9a3e005
--- /dev/null
+++ b/gerrit-server/src/main/resources/com/google/gerrit/server/mail/ChangeMoved.vm
@@ -0,0 +1,44 @@
+## Copyright (C) 2011 The Android Open Source Project
+##
+## Licensed under the Apache License, Version 2.0 (the "License");
+## you may not use this file except in compliance with the License.
+## You may obtain a copy of the License at
+##
+## http://www.apache.org/licenses/LICENSE-2.0
+##
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS,
+## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+## See the License for the specific language governing permissions and
+## limitations under the License.
+##
+##
+## Template Type:
+## -------------
+## This is a velocity mail template, see: http://velocity.apache.org and the
+## gerrit-docs:config-mail.txt for more info on modifying gerrit mail templates.
+##
+## Template File Names and extensions:
+## ----------------------------------
+## Gerrit will use templates ending in ".vm" but will ignore templates ending
+## in ".vm.example".  If a .vm template does not exist, the default internal
+## gerrit template which is the same as the .vm.example will be used.  If you
+## want to override the default template, copy the .vm.exmaple file to a .vm
+## file and edit it appropriately.
+##
+## This Template:
+## --------------
+## The Moved.vm template will determine the contents of the email related
+## to a change being moved to another destination branch.  It
+## is a ChangeEmail: see ChangeSubject.vm and ChangeFooter.vm.
+##
+$fromName has moved this change to a new destination branch: $change.dest.shortName.
+
+Change subject: $change.subject
+......................................................................
+
+
+#if ($coverLetter)
+$coverLetter
+
+#end
-- 
1.7.8

